packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts

```diff
@@ -0,0 +1,434 @@
+     */
+    prepareForEdit(): void;
+
+    /**
+     * Call to free resources.
+     * EditableTrees created in this context are invalid to use after this.
+     */
+    free(): void;
+}
+
+class ProxyContext implements EditableTreeContext {
+    public readonly withCursors: Set<ProxyTarget> = new Set();
+    public readonly withAnchors: Set<ProxyTarget> = new Set();
+    constructor(public readonly forest: IEditableForest) {}
+
+    public prepareForEdit(): void {
+        for (const target of this.withCursors) {
+            target.prepareForEdit();
+        }
+        assert(this.withCursors.size === 0, "prepareForEdit should remove all cursors");
+    }
+    public free(): void {
+        for (const target of this.withCursors) {
+            target.free();
+        }
+        for (const target of this.withAnchors) {
+            target.free();
+        }
+        assert(this.withCursors.size === 0, "free should remove all cursors");
+        assert(this.withAnchors.size === 0, "free should remove all anchors");
+    }
+}
+
+class ProxyTarget {
+	private readonly lazyCursor: ITreeSubscriptionCursor;
+    private anchor?: Anchor;
+
+	constructor(
+		public readonly context: ProxyContext,
+		cursor: ITreeSubscriptionCursor,
+	) {
+		this.lazyCursor = cursor.fork();
+        context.withCursors.add(this);
+	}
+
+    public free(): void {
+        this.lazyCursor.free();
+        this.context.withCursors.delete(this);
+        if (this.anchor !== undefined) {
+            this.context.forest.anchors.forget(this.anchor);
+            this.context.withAnchors.delete(this);
+            this.anchor = undefined;
+        }
+    }
+
+    public prepareForEdit(): void {
+        if (this.anchor === undefined) {
+            this.anchor = this.lazyCursor.buildAnchor();
+            this.context.withAnchors.add(this);
+        }
+        this.lazyCursor.clear();
+        this.context.withCursors.delete(this);
+    }
+
+	public get cursor(): ITreeSubscriptionCursor {
+		if (this.lazyCursor.state === ITreeSubscriptionCursorState.Cleared) {
+            assert(this.anchor !== undefined, "EditableTree should have an anchor if it does not have a cursor");
+			const result = this.context.forest.tryMoveCursorTo(this.anchor, this.lazyCursor);
+			assert(result === TreeNavigationResult.Ok,
+                "It is invalid to access an EditableTree node which no longer exists");
+            this.context.withCursors.add(this);
+            this.context.forest.anchors.forget(this.anchor);
+            this.context.withAnchors.delete(this);
+            this.anchor = undefined;
+		}
+		return this.lazyCursor;
+	}
+
+	get type(): TreeSchema {
+		return this.context.forest.schema.lookupTreeSchema(this.cursor.type);
+	}
+
+    get value(): Value {
+		return this.cursor.value;
+	}
+
+    public lookupFieldKind(key: string): FieldKind {
+        return getFieldKind(getFieldSchema(this.type, key));
+    }
+
+    get keys(): string[] {
```

Comment:
Because this is a getter, having it return a mutable array is confusing (a caller might expect modifying it to update the object). I suggest either making this a readonly array, or not making this a getter and using a normal method instead.